Added Instructions for using Helm using Job-Executor (#162)
keptn-contrib/artifacthub: helm/0.1.0/README.md

# Helm Job-Executor Integration

> This integration shows you how to leverage [Job-executor-service](https://github.com/keptn-contrib/job-executor-service) for deployment with `helm` on your Kubernetes cluster.

Reviewer: Are you saying that an integration can be just documentation?  That seems like a confusing use of the term "integration".

Author: Yes, that's what I am saying here. And we already have plenty of those (everything related to webhook-service and job-executor), so I'm not inventing the words here. Example: https://artifacthub.io/packages/keptn/keptn-integrations/githubaction

> **Please note, that the instructions provided here serve the purpose of enabling you to run and configure your Kubernetes deployments using job-executor-service. They are in no way intended to be a complete reference of Kubernetes, nor Helm, nor Keptn.**

Reviewer: ```suggestion
**Please note, that the instructions provided here enable you to run and configure your Kubernetes deployments using the job-executor-service. They are in no way intended to be a complete reference of Kubernetes, nor Helm, nor Keptn.**
```

## Installation and Configuration

### Step 1: Install the Job-Executor in your cluster

> Install [Job-Executor documentation](https://github.com/keptn-contrib/job-executor-service), and make sure it is subscribed to the following Keptn Cloud Events:

Reviewer: Do you really mean "Install the ... documentation"?

Author: fixed

* `sh.keptn.event.deployment.triggered` (for actual deployments)
* `sh.keptn.event.rollback.triggered` (for rollback sequence, i.e., when a previous deployment does not go well)
* `sh.keptn.event.action.triggered` (for remediation sequence)

**Example Installation Instruction for Job Executor**
```bash
TASK_SUBSCRIPTION='sh.keptn.event.deployment.triggered\,sh.keptn.event.rollback.triggered\,sh.keptn.event.action.triggered'
JES_VERSION=0.2.5-next.0
JES_NAMESPACE=keptn-jes

helm upgrade --install --create-namespace -n ${JES_NAMESPACE} \
  job-executor-service https://github.com/keptn-contrib/job-executor-service/releases/download/${JES_VERSION}/job-executor-service-${JES_VERSION}.tgz \
  --set remoteControlPlane.autoDetect.enabled="true",remoteControlPlane.topicSubscription="${TASK_SUBSCRIPTION}",remoteControlPlane.api.token="",remoteControlPlane.api.hostname="",remoteControlPlane.api.protocol=""
```

> **In case you have installed helm-service, please uninstall it**

Reviewer: ```suggestion
**If you have installed helm-service, please uninstall it**
```
Should this be done before or after you  install JES?

Author: Actually doesn't matter.

```bash
helm uninstall jmeter-service -n keptn
```

> ### Step 2: Make sure your project is setup properly with the right tasks

Reviewer: ```suggestion
### Step 2: Make sure your project is set up properly with the right tasks.
```

              properties:
                timeframe: "2m"

        - name: "rollback"
          triggeredOn:
            - event: "qa.delivery.finished"
              selector:
                match:
                  result: "fail"
          tasks:
            - name: "rollback"

    - name: "production"
      sequences:
        - name: "delivery"
          triggeredOn:
            - event: "qa.delivery.finished"
          tasks:
            - name: "approval"
              properties:
                pass: "manual"
                warning: "manual"
            - name: "deployment"

        - name: "remediation"
          triggeredOn:
            - event: "production.remediation.finished"
              selector:
                match:
                  evaluation.result: "fail"
          tasks:
            - name: "get-action"
            - name: "action"
            - name: "evaluation"
              triggeredAfter: "2m"
              properties:
                timeframe: "2m"

```

### Step 3: Add the Job-Executor configuration file

The following job-executor configuration file (referred to as job config or `job/config.yaml`) allows you to
* deploy using `helm upgrade`,

> * rollback using `helm rollback`, and

Reviewer: ```suggestion
* roll back using `helm rollback`, and
```

Author: "rollback" is actually a word commonly used in these scenarios (e.g., release management, databases, etc...), so I'd rather keep it :)

  - name: "Scale using kubectl"
    events:
      - name: "sh.keptn.event.action.triggered"
        jsonpath:
          property: "$.data.action.action"
          match: "scaling"
    tasks:
      - name: "Run kubectl"
        serviceAccount: "jes-deploy-using-helm"
        env:
          - name: SCALING
            value: $.data.action.value
            valueFrom: event
        image: "alpine/k8s:1.20.15"
        cmd: ["sh"]
        # Note: Hardcoded kubernetes namespace & KEPTN_SERVICE does most likely not match the deployment name
        args: ["-c", "REPLICAS=$(kubectl -n ${KEPTN_PROJECT}-${KEPTN_STAGE} get deployment/${KEPTN_SERVICE} -o go-template='{{.spec.replicas}}');DESIRED=$((${SCALING}+${REPLICAS}));echo Scaling deployment to ${DESIRED} && kubectl -n ${KEPTN_PROJECT}-${KEPTN_STAGE} scale --replicas=${DESIRED} deployment/${KEPTN_SERVICE}"]
```

Add the job-executor configuration file to your Keptn service and stage:

```bash
PROJECT=podtato-head
SERVICE=helloservice

keptn add-resource --project=$PROJECT --service=$SERVICE --all-stages --resource=helm-job-config.yaml --resourceUri=job/config.yaml
```

**Accessing the Kubernetes API**

> By default, job-executor-service does not grant access to the Kubernetes api, the respective jobs (e.g., `helm install` or `helm upgrade`) would fail with `The connection to the server localhost:8080 was refused - did you specify the right host or port?`.

Reviewer: ```suggestion
By default, `job-executor-service` does not grant access to the Kubernetes API, so the respective jobs (e.g., `helm install` or `helm upgrade`) would fail with `The connection to the server localhost:8080 was refused - did you specify the right host or port?`.
```
It seems to me that you should either say "the Job Executor Service", "JES" (assumes you have defined the acronym somewhere at the top of the file), or you should put tick marks around job-executor-service.

Author: Good suggestion, I will rework the document!

* `sh.keptn.event.action.triggered` (for remediation sequence)

This can verified in Keptn Bridge -> Project -> Settings -> Integrations -> job-executor-service.

![](https://raw.githubusercontent.com/keptn-sandbox/artifacthub/main/helm/0.1.0/images/bridge_subscriptions.png)

**Example Installation Instruction for Job Executor**
Please update `JES_VERSION` and `JES_NAMESPACE` in the example below according to your needs.

```bash
TASK_SUBSCRIPTION='sh.keptn.event.deployment.triggered\,sh.keptn.event.rollback.triggered\,sh.keptn.event.action.triggered'
JES_VERSION=0.2.5
JES_NAMESPACE=keptn-jes

helm upgrade --install --create-namespace -n ${JES_NAMESPACE} \
  job-executor-service https://github.com/keptn-contrib/job-executor-service/releases/download/${JES_VERSION}/job-executor-service-${JES_VERSION}.tgz \
  --set remoteControlPlane.autoDetect.enabled="true",remoteControlPlane.topicSubscription="${TASK_SUBSCRIPTION}",remoteControlPlane.api.token="",remoteControlPlane.api.hostname="",remoteControlPlane.api.protocol=""
```

**If you have installed helm-service, please uninstall it**
```bash

> helm uninstall jmeter-service -n keptn

Author: should be helm-service
